### DISCUSSION TITLE
We need an official mechanism to allow runtime configuration from process.env

### DISCUSSION THREAD

**Discussion by tomcollins** [vercel/next.js#52173] [App Router] [answered]
I'm new to Next and I'm really enjoying the app router and SSR. Great work.

I'm working through proving some "enterprise" requirements e.g. structured json logging (I have a slightly nasty solution that kinda works). 

I'm now working through runtime configuration. I want to deploy my app to kubernetes. There will be a different configuration in each env (dev, preprod, perf, prod etc). I need runtime configuration to work like any other application we deploy to k8s, where we can pass in a bunch of ENV vars when starting a pod. 

**Why not use build time config ?**
Build time configuration will not work in this sort of context, as the ENV config gets baked in at build time. I want to build a single Docker image that I can deploy to different environments, with different runtime config. Just like every other application we deploy to kubernetes.

I can't imagine I'm alone in wanting to do this. There are some [hacky workarounds](https://raphaelpralat.medium.com/system-environment-variables-in-next-js-with-docker-1f0754e04cde). There are [other related discussions](https://github.com/vercel/next.js/discussions/36338#discussioncomment-5767179) and [here](https://github.com/vercel/next.js/discussions/34894).

I may be missing some fundamental understanding of why Next appears to be moving away from this ([runtime config is considered legacy](https://nextjs.org/docs/app/api-reference/next-config-js/runtime-configuration)) with the App router, in which case I'd benefit from understanding this a little more.

**Comment by icyJoseph** [comment]
Hi,

I think what you want really is a page that's rendered on demand, `λ  (Server)  server-side renders at runtime (uses getInitialProps or getServerSideProps)` kind of thing, because otherwise, pre-rendering would totally inline values, given that the objective of pre-rendering is having a fully rendered, self standing, page.

And yeah, even if made on demand, the values are inlined into the source code. This is where we need to use `no-store` or `no-cache` based fetch, so that they opt the route into `λ` behaviour.

- https://github.com/vercel/next.js/discussions/51212
- https://github.com/vercel/next.js/discussions/49260

In short, the fetch would be made to fail, but instructed to not cache/store, and instead in a catch branch, we read from the `process.env`.

A pre-rendered page is tricker, and perhaps not doable. If you want to have runtime config, I think you probably need to render on demand as well. 

I guess, the framework could expose a function that's seen a "dynamic" like "cookies" or "headers", and using them should opt-in to dynamic rendering, there you should have access to the underlying process.env, in some way.

**Reply by tomcollins** [reply, marked as answer]
Hi @icyJoseph. Thanks for taking the time to answer this. It has helped me understand what Next is doing and how to achieve what I'm trying to do.

For clarity I'm quite happy with dynamic on-demand rendering. I'm not particularly concerned with pre-rending for my use case.

I've now managed to use runtime env variables passed into my Docker container on Kubernetes including AWS credentials including [IRSA](https://docs.aws.amazon.com/eks/latest/userguide/iam-roles-for-service-accounts.html).

### Server Vs Static renders

Understanding when this is happening via the `next build` output and what the triggers are has been key for me. Along with  learning about `export dynamic='force-dynamic'` ([docs](https://nextjs.org/docs/app/api-reference/file-conventions/route-segment-config)) which I was not aware of when asking the question. 

#### Example - Static Render

An API route handler located in /app/api/config/route.ts that outputs the value of process.env.

In the case the page is marked as ○ static in the build output. The value of `process.env` will be baked in at build time, go and look at the build output in `.next/server/app/api/config.body` if you're interested.

```typescript
import { NextResponse } from 'next/server'
 
export async function GET() {
    return NextResponse.json({
        env: process.env
    })
}
``` 

Output from `next build`

```bash
Route (app)                                Size     First Load JS
┌ ○ /                                      141 B          80.2 kB
└ ○ /api/config                            0 B                0 B

λ  (Server)  server-side renders at runtime (uses getInitialProps or getServerSideProps)
○  (Static)  automatically rendered as static HTML (uses no initial props)
```

#### Dynamic Render

In this example we add `export const dynamic = 'force-dynamic';` to the route handler. This results in the build process making the handler as λ dynamic,  `process.env will be evaluated at runtime and `.next/server/app/api/config.body` does not exist.

The updated handler:

```typescript
import { NextResponse } from 'next/server'

export const dynamic = 'force-dynamic';
 
export async function GET() {
    return NextResponse.json({
        env: process.env
    })
}
``` 

Output from `next build`

```bash
Route (app)                                Size     First Load JS
┌ ○ /                                      141 B          80.2 kB
└ λ /api/config                            0 B                0 B

λ  (Server)  server-side renders at runtime (uses getInitialProps or getServerSideProps)
○  (Static)  automatically rendered as static HTML (uses no initial props)
```

Happy to close this.

**Reply by jalessio** [reply]
@tomcollins This is an incredibly helpful writeup. Thank you!
